pkg/scheduler/predicates/ha.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicates
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb-operator/pkg/label"
+	"github.com/pingcap/tidb-operator/pkg/util"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+type ha struct {
+	kubeCli      kubernetes.Interface
+	podListFn    func(ns, clusterName, component string) (*apiv1.PodList, error)
+	pdReplicas   int32
+	tikvReplicas int32
+}
+
+// NewHA returns a Predicate
+func NewHA(kubeCli kubernetes.Interface, pdReplicas, tikvReplicas int32) Predicate {
+	h := &ha{
+		kubeCli:      kubeCli,
+		pdReplicas:   pdReplicas,
+		tikvReplicas: tikvReplicas,
+	}
+	h.podListFn = h.realPodListFn
+	return h
+}
+
+func (h *ha) Name() string {
+	return "HighAvailability"
+}
+
+func (h *ha) Filter(clusterName string, pod *apiv1.Pod, nodes []apiv1.Node) ([]apiv1.Node, error) {
+	ns := pod.GetNamespace()
+	podName := pod.GetName()
+
+	var component string
+	var exist bool
+	if component, exist = pod.Labels[label.ComponentLabelKey]; !exist {
+		return nil, fmt.Errorf("can't find component in pod labels: %s/%s", ns, podName)
+	}
+	podList, err := h.podListFn(ns, clusterName, component)
+	if err != nil {
+		return nil, err
+	}
+
+	ordinal, err := util.GetOrdinalFromPodName(podName)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeMap := make(map[string][]string)
+	for _, node := range nodes {
+		nodeMap[node.GetName()] = make([]string, 0)
+	}
+	for _, pod := range podList.Items {
+		podName1 := pod.GetName()
+		nodeName := pod.Spec.NodeName
+		ordinal1, err := util.GetOrdinalFromPodName(podName1)
+		if err != nil {
+			return nil, err
+		}
+
+		if ordinal1 < ordinal && nodeName == "" {
+			return nil, fmt.Errorf("waiting for pod: %s/%s to be scheduled", ns, podName1)
+		}
+		if nodeName == "" {
+			continue
+		}
+
+		nodeMap[nodeName] = append(nodeMap[nodeName], podName1)
```

Comment:
if the nodeName is not contain in nodeMap, there will happen panic ,we should add logic here
```
if nodeMap[nodeName] ==nil{
      continue
}
```

Reply:
Good catch!
Done!